buildpacks/metrics-agent/src/layers/install_agentmon.rs

```diff
@@ -0,0 +1,329 @@
+        LayerTypes {
+            build: true,
+            launch: true,
+            cache: true,
+        }
+    }
+
+    fn create(
+        &self,
+        _context: &libcnb::build::BuildContext<Self::Buildpack>,
+        layer_path: &std::path::Path,
+    ) -> Result<
+        libcnb::layer::LayerResult<Self::Metadata>,
+        <Self::Buildpack as libcnb::Buildpack>::Error,
+    > {
+        let destination_dir = layer_path.join("bin");
+        let executable = destination_dir.join("agentmon");
+
+        let mut timer = self.section.say_with_inline_timer("Downloading");
+
+        let url = get_latest_url()
+            .map_err(DownloadAgentmonError::CannotGetLatestUrl)
+            .map_err(MetricsAgentError::DownloadAgentmonError)?;
+
+        download_to_dir(&destination_dir, &url)
+            .map_err(MetricsAgentError::DownloadAgentmonError)?;
+        timer.done();
+
+        self.section.say("Writing scripts");
+        let execd = write_execd(&executable, layer_path)
+            .map_err(MetricsAgentError::DownloadAgentmonError)?;
+
+        LayerResultBuilder::new(Metadata {
+            download_url: Some(url),
+        })
+        .exec_d_program("spawn agentmon", execd)
+        .build()
+    }
+
+    fn existing_layer_strategy(
+        &self,
+        _context: &libcnb::build::BuildContext<Self::Buildpack>,
+        layer_data: &libcnb::layer::LayerData<Self::Metadata>,
+    ) -> Result<libcnb::layer::ExistingLayerStrategy, <Self::Buildpack as libcnb::Buildpack>::Error>
+    {
+        if let Some(old_url) = &layer_data.content_metadata.metadata.download_url {
+            let url = get_latest_url()
+                .map_err(DownloadAgentmonError::CannotGetLatestUrl)
+                .map_err(MetricsAgentError::DownloadAgentmonError)?;
+
+            if old_url == &url {
+                self.section.say("Using cache");
+                Ok(ExistingLayerStrategy::Keep)
+            } else {
+                let url = build_output::fmt::value(url);
+                self.section
+                    .say_with_details("Clearing cache", format!("Updated url {url}"));
+                Ok(ExistingLayerStrategy::Recreate)
+            }
+        } else {
+            self.section
+                .say_with_details("Clearing cache", "No url found in metadata");
+
+            Ok(ExistingLayerStrategy::Recreate)
+        }
+    }
+
+    fn migrate_incompatible_metadata(
+        &self,
+        _context: &libcnb::build::BuildContext<Self::Buildpack>,
+        _metadata: &GenericMetadata,
+    ) -> Result<
+        libcnb::layer::MetadataMigration<Self::Metadata>,
+        <Self::Buildpack as libcnb::Buildpack>::Error,
+    > {
+        self.section
+            .say_with_details("Clearing cache", "invalid metadata");
+
+        Ok(libcnb::layer::MetadataMigration::RecreateLayer)
+    }
+}
+
+fn write_execd(agentmon_path: &Path, layer_path: &Path) -> Result<PathBuf, DownloadAgentmonError> {
```

Comment:
`DownloadAgentmonError` seems wrong here. Either the name must change or the error should be specific to this function and wrapped at the call-site.